src/main/kotlin/org/rust/ide/typing/paste/RsImportCopyPasteProcessor.kt

```diff
@@ -0,0 +1,257 @@
+    }
+}
+
+class RsTextBlockTransferableData(val importMap: ImportMap) : TextBlockTransferableData, Cloneable {
+    override fun getFlavor(): DataFlavor? = RsImportCopyPasteProcessor.dataFlavor
+
+    override fun getOffsetCount(): Int = 0
+
+    override fun getOffsets(offsets: IntArray?, index: Int): Int = index
+    override fun setOffsets(offsets: IntArray?, index: Int): Int = index
+
+    public override fun clone(): RsTextBlockTransferableData {
+        try {
+            return super.clone() as RsTextBlockTransferableData
+        } catch (e: CloneNotSupportedException) {
+            throw RuntimeException()
+        }
+    }
+}
+
+class RsImportCopyPasteProcessor : CopyPastePostProcessor<RsTextBlockTransferableData>() {
+    override fun collectTransferableData(
+        file: PsiFile,
+        editor: Editor,
+        startOffsets: IntArray,
+        endOffsets: IntArray
+    ): List<RsTextBlockTransferableData> {
+        if (file !is RsFile || DumbService.getInstance(file.getProject()).isDumb) return emptyList()
+
+        val ranges = startOffsets.indices.map { TextRange(startOffsets[it], endOffsets[it]) }
+
+        val map = if (ranges.size == 1) {
+            createFqnMap(file, ranges[0])
+        } else {
+            ImportMap.Empty
+        }
+
+        return listOf(
+            RsTextBlockTransferableData(map)
+        )
+    }
+
+    override fun extractTransferableData(content: Transferable): List<RsTextBlockTransferableData> {
+        try {
+            val data = content.getTransferData(dataFlavor) as? RsTextBlockTransferableData ?: return emptyList()
+            // copy to prevent changing of original by convertLineSeparators
+            return listOf(data.clone())
+        } catch (ignored: UnsupportedFlavorException) {
+        } catch (ignored: IOException) {
+        }
+        return emptyList()
+    }
+
+    override fun processTransferableData(
+        project: Project,
+        editor: Editor,
+        bounds: RangeMarker,
+        caretOffset: Int,
+        indented: Ref<in Boolean>,
+        values: List<RsTextBlockTransferableData>
+    ) {
+        PsiDocumentManager.getInstance(project).commitAllDocuments()
+
+        val data = values.getOrNull(0) ?: return
+        val rsFile = editor.document.toPsiFile(project) as? RsFile ?: return
+        val range = bounds.range
+
+        val elements = gatherElements(rsFile, range)
+        val importCtx = elements.firstOrNull { it is RsElement } as? RsElement ?: return
+
+        val visitor = ImportingVisitor(importCtx, range, data)
+
+        runWriteAction {
+            for (element in elements) {
+                element.accept(visitor)
+            }
+            // We need to import the candidates after visiting all elements, otherwise the relative offsets could be
+            // invalidated after an import has been added
+            for ((candidate, ctx) in visitor.importCandidates) {
+                candidate.import(ctx)
```

Comment:
Looks like `ctx` always equals to `importCtx` so we can use it directly

```suggestion
                candidate.import(importCtx)
```